src/hmem_cuda.c

```diff
@@ -482,6 +493,46 @@ static int cuda_hmem_verify_devices(void)
 	return FI_SUCCESS;
 }
 
+/*
+ * Loop over pairs of available devices to deduce the overall P2P capabilities
+ */
+static int cuda_hmem_detect_p2p_access_support(void)
```

Comment:
I am bit concerned this loop will considerably increase the initialization time cost.

Reply:
Tested this with 2 MPI processes on a single node with a g4dn.12xlarge (8 GPUs) and altered the code to let the full loop execute, and every log statement after calling `cuDeviceGetP2PAttribute()` (112 total calls) had the exact same timestamp - indicating little initialization overhead.

Can't say how it scales in practice beyond 8 GPUs per node, but this looks perfectly fine to me.

Reply:
I am curious to see how it impacts the performance for more ranks per node. But I do not think having more ranks per node could change the story much since this is only called once during initialization.

Reply:
But can we just simply check 1 pair? This loop looks a bit over-engineering to me.

Reply:
> But can we just simply check 1 pair? This loop looks a bit over-engineering to me.

I'd be more comfortable checking peer access between GPU 0 and the rest of the CUDA devices, which reduces the time complexity to $O(n)$. Checking a single pair doesn't feel bulletproof enough for me.

Reply:
O(n) is better than O(n^2) to me

Reply:
Actually on second thought, this should be just as effective as the current $O(n^2)$ search:

```c
for (dev = 0, peer = 1; peer < cuda_attr.device_count; dev += 2, peer += 2) {
    ...
}
```

This way each GPU is checked only once

Reply:
your new revision is actually `++dev, ++peer`?

Reply:
Fair enough.

Reply:
I'm not familiar with the IPC limitations of CUDA devices but is it possible that 0->1 is ok, 1->2 is ok, but 0->2 is not?

Reply:
> I'm not familiar with the IPC limitations of CUDA devices but is it possible that 0->1 is ok, 1->2 is ok, but 0->2 is not?

I don't believe so. In this scenario, if 0 cannot access 2, that means 0 cannot access 1 and/or 1 cannot access 2. This would indicate some kind of mixed topology/configuration. It would be recommended to restrict the devices visible to Libfabric via `CUDA_VISIBLE_DEVICES` to only "similar" devices.

On that note, this makes me realize we definitely shouldn't employ the [(0,1),(2,3),...] check since that wouldn't catch potential "dissimilar" devices e.g. (1,2). So I think the simpler linear check is perfectly suitable here.

Reply:
I think I would have pinned dev=0 and perform (0-1, 0-2, 0-3 ...), seems clearer to me.  My 2c.